packages/profiling-node/src/integration.ts

```diff
@@ -24,8 +31,296 @@ function addToProfileQueue(profile: RawThreadCpuProfile): void {
+    }
+
+    const profile_id = maybeProfileSpan(client, span, undefined);
+
+    if (profile_id) {
+      const options = client.getOptions();
+      // Not intended for external use, hence missing types, but we want to profile a couple of things at Sentry that
+      // currently exceed the default timeout set by the SDKs.
+      const maxProfileDurationMs =
+        (options._experiments && options._experiments['maxProfileDurationMs']) || MAX_PROFILE_DURATION_MS;
+
+      if (PROFILE_TIMEOUTS[profile_id]) {
+        global.clearTimeout(PROFILE_TIMEOUTS[profile_id]);
+        // eslint-disable-next-line @typescript-eslint/no-dynamic-delete
+        delete PROFILE_TIMEOUTS[profile_id];
+      }
+
+      // Enqueue a timeout to prevent profiles from running over max duration.
+      const timeout = global.setTimeout(() => {
+        DEBUG_BUILD &&
+          logger.log('[Profiling] max profile duration elapsed, stopping profiling for:', spanToJSON(span).description);
+
+        const profile = stopSpanProfile(span, profile_id);
+        if (profile) {
+          addToProfileQueue(profile);
+        }
+      }, maxProfileDurationMs);
+
+      // Unref timeout so it doesn't keep the process alive.
+      timeout.unref();
+
+      getCurrentScope().setContext('profile', { profile_id });
+      spanToProfileIdMap.set(span, profile_id);
+    }
+  });
+
+  client.on('spanEnd', span => {
+    const profile_id = spanToProfileIdMap.get(span);
+
+    if (profile_id) {
+      if (PROFILE_TIMEOUTS[profile_id]) {
+        global.clearTimeout(PROFILE_TIMEOUTS[profile_id]);
+        // eslint-disable-next-line @typescript-eslint/no-dynamic-delete
+        delete PROFILE_TIMEOUTS[profile_id];
+      }
+      const profile = stopSpanProfile(span, profile_id);
+
+      if (profile) {
+        addToProfileQueue(profile);
+      }
+    }
+  });
+
+  client.on('beforeEnvelope', (envelope): void => {
+    // if not profiles are in queue, there is nothing to add to the envelope.
+    if (!PROFILE_QUEUE.length) {
+      return;
+    }
+
+    const profiledTransactionEvents = findProfiledTransactionsFromEnvelope(envelope);
+    if (!profiledTransactionEvents.length) {
+      return;
+    }
+
+    const profilesToAddToEnvelope: Profile[] = [];
+
+    for (const profiledTransaction of profiledTransactionEvents) {
+      const profileContext = profiledTransaction.contexts?.['profile'];
+      const profile_id = profileContext?.['profile_id'];
+
+      if (!profile_id) {
+        throw new TypeError('[Profiling] cannot find profile for a transaction without a profile context');
+      }
+
+      // Remove the profile from the transaction context before sending, relay will take care of the rest.
+      if (profileContext) {
+        delete profiledTransaction.contexts?.['profile'];
+      }
+
+      // We need to find both a profile and a transaction event for the same profile_id.
+      const profileIndex = PROFILE_QUEUE.findIndex(p => p.profile_id === profile_id);
```

Comment:
I should have posted that this was just copied over. It's been a while, but iirc it is because we treat this as a very simple ring buffer by just slicing it if the number is exceeded. Since we keep a small number of profiles in the queue, it is probably not an issue